### PR TITLE
Updated get weights to return only filter type layer weights

### PR DESCRIPTION
**Pull request by peacekurella**
* `get_weights` function now returns weights for layers of type Linear, Conv2D, or ConvTranspose2D or their derivatives.

* Tested with the classes `LitMLP` and `LitCNN`, added batch norm to both models, ensured that `filterlogcallback` functions as expected.

https://wandb.ai/wandb/lit-mlp-dev?workspace=user-prashanthkurella

### CODE REVIEW

File: callbacks.py [hunk 1]
```diff
@@ -317,17 +319,50 @@ def preds_from_y_hats(y_hats):
...
 
 
+def allowed_filter_type(module):
+    expr = isinstance(module, torch.nn.Conv2d)
+    expr = expr or issubclass(type(module), torch.nn.Conv2d)
+    expr = expr or isinstance(module, torch.nn.ConvTranspose2d)
+    expr = expr or issubclass(type(module), torch.nn.ConvTranspose2d)
+    expr = expr or isinstance(module, torch.nn.Linear)
+    expr = expr or issubclass(type(module), torch.nn.Linear)
+    expr = expr or isinstance(module, FullyConnected)
```
Comment (charlesfrye) [inline callbacks.py:361]: Since these modules contain `torch.nn.Linear` modules, there's no need to check them separately. I believe this prevents the need for the import of the `nn` Python submodule.

File: callbacks.py [hunk 2]
```diff
@@ -317,17 +319,50 @@ def preds_from_y_hats(y_hats):
...
-               if "weight" in name.split(".")[-1]]
-    masks = [mask for name, mask in module.named_buffers()
-             if "weight_mask" in name.split(".")[-1]]
+    weights = []
+    masks = []
+    for submodule in module.children():
+        # recursively go down to individual layers
+        if isinstance(submodule, torch.nn.Sequential):
+            parent = submodule
+            while(isinstance(submodule, torch.nn.Sequential)):
```
Comment (charlesfrye) [inline callbacks.py:328]: Style: I wouldn't use parens here for the `while`. While it's allowed in Python for any expression to have parentheses around it, they're generally only used to build tuples/call functions.

File: callbacks.py [hunk 3]
```diff
@@ -317,17 +319,50 @@ def preds_from_y_hats(y_hats):
 
 
 def get_weights(module):
-    weights = [parameter for name, parameter in module.named_parameters()
-               if "weight" in name.split(".")[-1]]
-    masks = [mask for name, mask in module.named_buffers()
-             if "weight_mask" in name.split(".")[-1]]
+    weights = []
```
Comment (charlesfrye) [inline callbacks.py:322]: Minor style: save a line without too much readability penalty with

```python
weights, masks = [], []
```

File: callbacks.py [hunk 4]
```diff
@@ -317,17 +319,50 @@ def preds_from_y_hats(y_hats):
...
+        elif allowed_filter_type(submodule):
+            for name, parameter in submodule.named_parameters():
+                if "weight" in name.split(".")[-1]:
+                    weights.append(parameter)
+            for name, mask in submodule.named_buffers():
+                if "weight_mask" in name.split(".")[-1]:
+                    masks.append(mask)
     if masks:
         with torch.no_grad():
             weights = [mask * weight for mask, weight in zip(masks, weights)]
```
Comment (charlesfrye) [inline callbacks.py:349]: Oof, just realized that this assumes that every weight has a mask, which is not always the case. All of the examples in the utils use it.

Could you pull the mask application logic out into a separate function (or check `torch.nn.prune` for an implementation!) and then use that function _before_ parameters get appended to the list? Make sure there's `no_grad`.

File: callbacks.py [hunk 5]
```diff
@@ -317,17 +319,50 @@ def preds_from_y_hats(y_hats):
 
 
 def get_weights(module):
```
Comment (charlesfrye) [inline callbacks.py:321]: This module traversal looks reasonably correct to me, but I actually think there's a better move: [`.modules`](https://discuss.pytorch.org/t/module-children-vs-module-modules/4551/4) travels recursively for you, and I think we could just iterate through the returned generator without needing to worry about recursing ourselves.

**Review by charlesfrye — CHANGES_REQUESTED** [review]
This looks correct, but I think there may be a better way (see comment on L321 about `.modules`) and I have some stylistic/readability improvements.
